Event Hubs: Update EventProcessorClient component to pull BlobClient from DI

oising opened the pull request:
- Update AzureComponent to pass concrete AzureClientFactoryBuilder and remove TBuilder generic
- Update associated components (RegisterClientFactory is replaced with concrete AddClient call)
- Update launchSettings.json in EH sample AppHost
- Update EventProcessorClient component to pull BlobServiceClient from DI

The BlobServiceClient is first looked for as a keyed service using `settings.BlobConnectionName` -- the `serviceKey` is always the same as the `connectionName` for keyed services. If this is not found, a default service is looked for. If this is not found, an `InvalidOperationException` is thrown with a helpful message. This is a breaking change from Preview 5.

Closes https://github.com/dotnet/aspire/issues/3254
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/3293)

oising committed:
update sample EH project launchsettings; add blob component for processor

oising committed:
update azurecomponent to expose serviceprovider; update associated components; update processor for DI retrieval of blob client

oising committed:
Merge branch 'main' into eventhubprocessor-blobclient-di

oising committed:
throw if blobclient cannot be retrieved from DI (this is a breaking change from P5)

oising committed:
fix tests

oising commented:
fyi @eerhardt

sebastienros commented on src/Components/Aspire.Azure.Data.Tables/AspireTablesExtensions.cs:
```diff
@@ -114,7 +116,7 @@ private sealed class TableServiceComponent : AzureComponent<AzureDataTablesSetti
                 return !string.IsNullOrEmpty(connectionString) ? new TableServiceClient(connectionString, options) :
                     cred is not null ? new TableServiceClient(settings.ServiceUri, cred, options) :
                     new TableServiceClient(settings.ServiceUri, options);
-            }, requiresCredential: false);
```
What is the impact of changing this? I see that it's always `true` in `AddClient` methods

eerhardt replied:
We shouldn't be changing this. We need to keep the current behavior.

eerhardt replied:
I fixed most changes back to the way they were. I missed the EventHubs ones, but I need to step away. Can you take care of that? I can take care of it when I get back too.

eerhardt committed:
PR feedback

sebastienros committed:
More PR feedback

sebastienros commented on src/Components/Aspire.Azure.Messaging.EventHubs/EventProcessorClientComponent.cs:
```diff
@@ -3,58 +3,62 @@
 
                 return processor;
-
-            }, requiresCredential: false);
```
There is no available method that provides options, cred, provider and requiresCredential. The ultimate private call has these but there is no public combination that allows to access it. Ideally on AddClient since this is the one we wanted. I assume the expectation was that if you provide a factory with `TokenCredential` you'd want `requiresCredential: true` but here we don't always.

eerhardt replied:
I was able to successfully use this using a connection string. So I'm not sure we need to specify `requiresCredential: false`. Since this is a new component, we shouldn't be breaking anyone. But let's keep the other components doing what they are doing now. This one can use AddClient.

@tg-msft - do you know what this `requiresCredential` bool does/means exactly? Why can't I specify `requiresCredential: false` when I use the overloads with IServiceProvider?

oising replied:
Yeah, this is the problem I had -- there was no public overload available that allowed specifying it when you want the service provider. I couldn't immediately see what the point of this flag even was, so I figured it was vestigial as everything seemed to work regardless.

tg-msft replied:
[requiresCredential](https://learn.microsoft.com/en-us/dotnet/api/azure.core.extensions.iazureclientfactorybuilderwithcredential.registerclientfactory?view=azure-dotnet#parameters) is a niche feature for services that allow anonymous auth (i.e., reads on a Storage account used for hosting static websites).  I believe it only controls whether or not we throw an exception when resolving a client if there was no credential provided.  https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/extensions/Microsoft.Extensions.Azure/src/Internal/ClientRegistration.cs#L46

eerhardt committed:
Merge remote-tracking branch 'upstream/main' into eventhubprocessor-blobclient-di

eerhardt committed:
Rename BlobClientConnectionName to BlobClientServiceKey.

Remove the need to specify a BlobClientServiceKey, and use an unkeyed client if none is provided.

davidfowl commented on src/Components/Aspire.Azure.Messaging.EventHubs/AzureMessagingEventHubsSettings.cs:
```diff
@@ -120,17 +120,20 @@ public sealed class AzureMessagingEventHubsConsumerSettings : AzureMessagingEven
+    /// an un-keyed BlobServiceClient will be retrieved from the IServiceProvider. If a BlobServiceClient is not available in
+    /// the IServiceProvider, an exception is thrown.
+    /// </remarks>
+    public string? BlobClientServiceKey{ get; set; }
```
Formatting

eerhardt committed:
fix tests

eerhardt reviewed (approved):
This looks great. I pushed a few minor changes.

Thanks for the great contributions @oising!

eerhardt committed:
fixup formatting
